plugins/inputs/dmcache/dmcache_linux.go

```diff
@@ -0,0 +1,131 @@
+// +build linux
+
+package dmcache
+
+import (
+	"strconv"
+	"strings"
+
+	"errors"
+
+	"github.com/influxdata/telegraf"
+)
+
+const metricName = "dmcache"
+
+func (c *DMCache) Gather(acc telegraf.Accumulator) error {
+	outputLines, err := c.getCurrentStatus()
+	if err != nil {
+		return err
+	}
+
+	total := make(map[string]interface{})
+
+	for _, s := range outputLines {
+		fields, err := parseDMSetupStatus(s)
+		if err != nil {
+			return err
+		}
+
+		if c.PerDevice {
+			tags := map[string]string{"device": fields["device"].(string)}
+			acc.AddFields(metricName, fields, tags)
+		}
+		aggregateStats(total, fields)
+	}
+
+	acc.AddFields(metricName, total, map[string]string{"device": "all"})
+
+	return nil
+}
+
+func parseDMSetupStatus(line string) (map[string]interface{}, error) {
+	var err error
+	parseError := errors.New("Output from dmsetup could not be parsed")
+	status := make(map[string]interface{})
+	values := strings.Fields(line)
+	if len(values) < 15 {
+		return nil, parseError
+	}
+
+	status["device"] = strings.TrimRight(values[0], ":")
```

Comment:
Still want these in a struct, this separates the responsibilities.  Call it something appropriate that represents the data, maybe CacheStatus.  I get why you did this but I think it will clear up some weirdness in the code, such as taking `device` from fields to add to tags.  You won't need to do type conversions and summing will be more explicit.  This open PR does a good job. #2031